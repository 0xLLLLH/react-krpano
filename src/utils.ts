import escapeHTML from 'escape-html';

type FuncName = 'set' | 'loadxml' | 'loadscene';

/**
 * 执行单个函数
 * @param func 函数名
 * @param params 参数列表
 */
export const buildKrpanoAction = (func: FuncName, params: Array<string | number | boolean>): string =>
    `${func}(${params.map((p) => p.toString()).join(', ')});`;

/**
 * 动态添加标签
 * @see https://krpano.com/forum/wbb/index.php?page=Thread&threadID=15873
 */
export const buildKrpanoTagSetterActions = (
    tag: 'scene' | 'hotspot' | 'layer',
    name: string,
    attrs: Record<string, string | boolean | number>,
): string =>
    Object.keys(attrs)
        .map((key) => {
            // 如果属性值中有双引号，需要改用单引号
            let quete = '"';
            if (attrs[key].toString().includes(quete)) {
                // eslint-disable-next-line quotes
                quete = "'";
            }
            if (key === 'style') {
                return `assignstyle(${tag}[${name}], ${attrs[key]});`;
            }
            if (typeof attrs[key] === 'boolean') {
                return `set(${tag}[${name}].${key}, ${attrs[key]});`;
            }
            // content是XML文本，不能转义，因为不涉及用户输入也不需要
            return `set(${tag}[${name}].${key}, ${quete}${
                key === 'content' ? attrs[key] : escapeHTML(attrs[key].toString())
            }${quete});`;
        })
        .join('');

export const Logger = {
    log: (...args: any[]): void => {
        if (process.env.NODE_ENV === 'development') {
            console.log(...args);
        }
    },
};

export interface XMLMeta {
    tag: string;
    attrs: Record<string, string | number | boolean>;
    children?: XMLMeta[];
}

/**
 * 根据元数据构建xml
 */
export const buildXML = ({ tag, attrs, children }: XMLMeta): string => {
    const attributes = Object.keys(attrs)
        .map((key) => `${key.toLowerCase()}="${attrs[key]}"`)
        .join(' ');

    if (children && children.length) {
        return `<${tag} ${attributes}>${children.map((child) => buildXML(child)).join('')}</${tag}>`;
    }
    return `<${tag} ${attributes} />`;
};
