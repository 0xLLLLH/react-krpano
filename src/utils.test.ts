import { buildKrpanoAction, buildKrpanoTagSetterActions, buildXML, mapObject, mapEventPropsToJSCall } from './utils';

describe('buildKrpanoAction()', () => {
    test('call', () => {
        expect(buildKrpanoAction('set', 'view.fov', 120)).toMatchSnapshot();

        expect(buildKrpanoAction('addhotspot', 'hotspotName')).toMatchSnapshot();

        expect(buildKrpanoAction('loadscene', 'scene1', 'null', 'MERGE', 'BLEND(0.5)')).toMatchSnapshot();
    });
});

describe('buildKrpanoTagSetterActions()', () => {
    test('call without tag name', () => {
        expect(buildKrpanoTagSetterActions('view', { fov: 90 })).toMatchSnapshot();
    });
    test('call with tag name', () => {
        expect(
            buildKrpanoTagSetterActions('scene[scene1]', { autoload: false, content: '<preview url="/preview.jpg" />' })
        ).toMatchSnapshot();

        expect(
            buildKrpanoTagSetterActions('hotspot[hotspotName]', {
                ath: 10,
                atv: 20,
                c: undefined,
            })
        ).toMatchSnapshot();
    });

    test('assignstyle', () => {
        expect(
            buildKrpanoTagSetterActions('hotspot[hotspotName]', {
                ath: 10,
                atv: 20,
                style: 'hotspot_style1',
            })
        ).toMatchSnapshot();
    });

    test('escape string', () => {
        expect(
            buildKrpanoTagSetterActions('hotspot[hotspotName]', {
                needEscape: `<div>It's normal string</div>`,
                content: `<xml>It's xml string, keep everything</xml>`,
            })
        ).toMatchSnapshot();
    });
});

describe('buildXML()', () => {
    test('empty attrs & children', () => {
        expect(
            buildXML({
                tag: 'tagName',
                attrs: {},
            })
        ).toMatchSnapshot();
    });

    test('empty children', () => {
        expect(
            buildXML({
                tag: 'tagName',
                attrs: {
                    key1: 'val1',
                    key2: 2,
                },
            })
        ).toMatchSnapshot();
    });

    test('children empty array', () => {
        expect(
            buildXML({
                tag: 'tagName',
                attrs: {
                    key1: 'val1',
                    key2: 2,
                },
                children: [],
            })
        ).toMatchSnapshot();
    });

    test('one childreny', () => {
        expect(
            buildXML({
                tag: 'tagName',
                attrs: {
                    key1: 'val1',
                    key2: 2,
                },
                children: [
                    {
                        tag: 'inner',
                        attrs: {
                            k: 'v',
                        },
                    },
                ],
            })
        ).toMatchSnapshot();
    });
});

describe('mapObject()', () => {
    test('call', () => {
        const obj = { a: 1, b: 2 };

        expect(mapObject(obj, (k, v) => ({ [k]: (v as number) + 1 }))).toEqual({
            a: 2,
            b: 3,
        });
    });
});

describe('mapEventPropsToJSCall()', () => {
    test('call', () => {
        const obj = { a: 1, b: 2, onClick: () => console.log(1) };

        expect(mapEventPropsToJSCall(obj, k => `js(${k}, selector)`)).toEqual({
            onClick: 'js(onClick, selector)',
        });
    });
});
