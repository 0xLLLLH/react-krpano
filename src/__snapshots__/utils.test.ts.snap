// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildKrpanoAction() call 1`] = `"set(view.fov, 120);"`;

exports[`buildKrpanoAction() call 2`] = `"addhotspot(hotspotName);"`;

exports[`buildKrpanoAction() call 3`] = `"loadscene(scene1, null, MERGE, BLEND(0.5));"`;

exports[`buildKrpanoTagSetterActions() assignstyle 1`] = `"set(hotspot[hotspotName].ath, 10);set(hotspot[hotspotName].atv, 20);assignstyle(hotspot[hotspotName], hotspot_style1);"`;

exports[`buildKrpanoTagSetterActions() call with tag name 1`] = `"set(scene[scene1].autoload, false);set(scene[scene1].content, '<preview url=\\"/preview.jpg\\" />');"`;

exports[`buildKrpanoTagSetterActions() call with tag name 2`] = `"set(hotspot[hotspotName].ath, 10);set(hotspot[hotspotName].atv, 20);"`;

exports[`buildKrpanoTagSetterActions() call without tag name 1`] = `"set(view.fov, 90);"`;

exports[`buildKrpanoTagSetterActions() escape string 1`] = `"set(hotspot[hotspotName].needescape, \\"&lt;div&gt;It&#39;s normal string&lt;/div&gt;\\");set(hotspot[hotspotName].content, \\"<xml>It's xml string, keep everything</xml>\\");"`;

exports[`buildXML() children empty array 1`] = `"<tagName key1=\\"val1\\" key2=\\"2\\" />"`;

exports[`buildXML() empty attrs & children 1`] = `"<tagName  />"`;

exports[`buildXML() empty children 1`] = `"<tagName key1=\\"val1\\" key2=\\"2\\" />"`;

exports[`buildXML() one childreny 1`] = `"<tagName key1=\\"val1\\" key2=\\"2\\"><inner k=\\"v\\" /></tagName>"`;
